.yarn/patches/@metamask-assets-controllers-npm-38.2.0-40af2afaa7.patch

```diff
@@ -0,0 +1,36 @@
+diff --git a/dist/assetsUtil.cjs b/dist/assetsUtil.cjs
+index e90a1b6767bc8ac54b7a4d580035cf5db6861dca..2662df7869e9079f9379c4dd1ded40b5af4e71f2 100644
+--- a/dist/assetsUtil.cjs
++++ b/dist/assetsUtil.cjs
+@@ -3,6 +3,7 @@ var __importDefault = (this && this.__importDefault) || function (mod) {
+     return (mod && mod.__esModule) ? mod : { "default": mod };
+ };
+ Object.defineProperty(exports, "__esModule", { value: true });
++function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { newObj[key] = obj[key]; } } } newObj.default = obj; return newObj; } }
+ exports.fetchTokenContractExchangeRates = exports.reduceInBatchesSerially = exports.divideIntoBatches = exports.ethersBigNumberToBN = exports.addUrlProtocolPrefix = exports.getFormattedIpfsUrl = exports.getIpfsCIDv1AndPath = exports.removeIpfsProtocolPrefix = exports.isTokenListSupportedForNetwork = exports.isTokenDetectionSupportedForNetwork = exports.SupportedTokenDetectionNetworks = exports.formatIconUrlWithProxy = exports.formatAggregatorNames = exports.hasNewCollectionFields = exports.compareNftMetadata = exports.TOKEN_PRICES_BATCH_SIZE = void 0;
+ const controller_utils_1 = require("@metamask/controller-utils");
+ const utils_1 = require("@metamask/utils");
+@@ -221,7 +222,7 @@ async function getIpfsCIDv1AndPath(ipfsUrl) {
+     const index = url.indexOf('/');
+     const cid = index !== -1 ? url.substring(0, index) : url;
+     const path = index !== -1 ? url.substring(index) : undefined;
+-    const { CID } = await import("multiformats");
++    const { CID } = await Promise.resolve().then(() => _interopRequireWildcard(require("multiformats")));
```

Comment:
The `await import()` call being left in there was intentional, because `await import()` is available to commonjs in node v12+, and passing an ESM-only module like `multiformats` into `require` crashes node or otherwise may cause undefined behavior.

The `import` call causing test runs to fail is also expected, because dynamic import is only available behind a `NODE_OPTIONS=--experimental-vm-modules` feature flag. This is why we've added that to all of our core test scripts (https://github.com/MetaMask/core/blob/main/packages/assets-controllers/package.json#L44-L47).

I guess the question here is whether `_interopRequireWildcard` is actually executing the `CID` import correctly, or if it's masking broken behavior.

Reply:
@MajorLift thnx a lot for looking into this 🙏  I have added some logs with this current patch and it looks like its working fine, its able to fetch CID and display the NFT media;

Without the patch, i can see the error `__import__ is not defined` in console again and its not displaying the images correctly; 
<img width="847" alt="Screenshot 2024-10-07 at 19 05 08" src="https://github.com/user-attachments/assets/f1c85a96-b744-4bb9-ae64-30a6936f1261">


Thanks for referencing the related test script update on [core](https://github.com/MetaMask/core/blob/main/packages/assets-controllers/package.json#L44-L47) 🙏  this is not  only related to tests, correct? since im able to see the error when running the app locally?

Reply:
The browser uses only ESM, so that's a really weird error message. Maybe it's coming from selenium or some other node process? I tried adding the flag to the e2e test build scripts (https://github.com/MetaMask/metamask-extension/pull/27675) but I'm still getting the same error. 

Since you verified that the import is working correctly I think we can move forward with the current patch. Thanks for being on top of that.

Would be nice to figure this error out for the future though. I'll do some more digging.

Reply:
This code is running under lavamoat and dynamic import is deliberately not supported - hence a transform changing all functions named `import` into `__import__`

We currently can't have dynamic import work inside compartments. It's not feasible to intercept it the way we'd need to. Since the bundler we use doesn't support it, the dynamic import would be called directly in the browser and "multiformats" is not a URL to a script file, so it would fail with a different error even if LavaMoat transforms weren't there to prevent it.
I recommend addressing this somehow in a minor refactor of the assets controller if it's intended for use in MetaMask.

When we switch to webpack with LavaMoat webpack plugin it's possible webpack will transform the dynamic import into its own syntax if configured for that, but that's the only hope for using dynamic require I see for now. Didn't check how that behaves yet. I can get back to you later.

Reply:
@naugtur Thanks for the insight! I wasn't aware that Lavamoat did this. I'll have to look into what else it blocks.

Unfortunately this a language-level CJS/ESM interop issue that isn't specific to `assets-controllers` and `multiformats` -- it applies to any ESM-only module in our dependency tree. 

Most of our modules are natively CJS. We use ESM syntax because we write in TypeScript,  but that is transpiled down to CJS in our published builds. Because of this, the only way for our modules to handle ESM imports is with dynamic import syntax i.e. `require` doesn't work, and neither does static import syntax that transpiles down to `require`.

> True CommonJS modules can require an ESM-transpiled-to-CJS module, since they’re both CommonJS at runtime. But in Node.js, require crashes if it resolves to an ES module. This means published libraries cannot migrate from transpiled modules to true ESM without breaking their CommonJS (true or transpiled) consumers:
```ts
// @Filename: node_modules/dependency/index.js
export function doSomething() { /* ... */ }
// @Filename: dependent.js
import { doSomething } from "dependency";
// ✅ Works if dependent and dependency are both transpiled
// ✅ Works if dependent and dependency are both true ESM
// ✅ Works if dependent is true ESM and dependency is transpiled
// 💥 Crashes if dependent is transpiled and dependency is true ESM
```
> https://www.typescriptlang.org/docs/handbook/modules/appendices/esm-cjs-interop.html#cannot-require-a-true-es-module

This issue was previously masked by our usage of the outdated `{ module: "commonjs", moduleResolution: "node10" }` tsconfig options. However, the core and snaps repos recently updated to `"node16"` for both options, which enforces sound module resolution restrictions and therefore prohibits ESM-only modules from being passed into `require` calls.

In this specific case, we can use the patch as is because `multiformats` apparently happens to work with `require` calls without breaking things, but this is a coincidence we can't rely on.

To preserve the ban on dynamic import syntax I see three possibilities:
1. Remove all ESM-only modules from all of our dependency trees, or replace them with CJS-compatible forks (we've done this before e.g. `@metamask/superstruct`).
2. Convert all of our modules to "true ESM".
3. Start using a bundler like Webpack in all of our repos to generate builds.

All of these would be major, if not prohibitive, undertakings.

Reply:
1 and 2 are not options we're fond of.
With the exception of: until we have a lavamoat-node that supports ESM (we're working on it) we might sometimes have to eliminate esm-only packages from builds that run under lavamoat.

We should be able to switch to webpack in the foreseeable future. LavaMoat Webpack plugin is reaching 1.0 by the end of this year and we're wery close to getting a working MetaMask build out of it. 3rdparty audit is finishing too.

Meanwhile, 
Joyee has made it possible to `require(esm)` in Node.js 
https://nodejs.org/api/esm.html#interoperability-with-commonjs
https://joyeecheung.github.io/blog/2024/03/18/require-esm-in-node-js/

So we're looking for a solution to the typescript situation that we can live with for limited time.

Reply:
Agreed that 1, 2 aren't the most ideal options. ;)

I wasn't aware of `--experimental-require-module`, but it looks like a great stopgap. Fingers crossed that we don't run into an ESM package with top-level `await`.

Based on anecdotal experience from applying `Node16` to core and snaps, our usage of CJS-incompatible packages is uncommon enough that we should be able to handle them on a case-by-case basis, at least for the time being. 

However, the few exceptions can have an outsized impact. Migrating from `superstruct` to `@metamask/superstruct` ended up being a [massive](https://github.com/MetaMask/snaps/pull/2445) [undertaking](https://github.com/MetaMask/core/pull/3645) because it required us to update and release basically all transitive dependents. Removing such a package will probably be just as difficult. Hopefully we don't come across such a scenario again any time soon.

Anyway, it's very good to know that there are fundamental fixes on the way, and thank you for your work on that! Especially excited to potentially have lavamoat-node directly applied to our upstream libraries.